src/System.CommandLine/Argument.cs

```diff
@@ -116,9 +116,20 @@ bool DefaultConvert(SymbolResult symbol, out object value)
 
         public object GetDefaultValue() => _defaultValue?.Invoke();
 
-        public void SetDefaultValue(object value) => SetDefaultValue(() => value);
+        public void SetDefaultValue(object value)
+        {
+            if (value is null)
```

Comment:
So I don't get this block here, since this implies calling Argument.SetDefaultValue((object)null) would throw, but Argument.SetDefaultValue(() => null) _would_ be fine, even though I'd expect those to meant the same thing.

Reply:
Agreed. I think the API needs to change so that intention is more clear between:

* there is a default value, and it is null, and
* there is no default value.

The latter is not currently representable in the API.

Reply:
In the latest commit, I've differentiated `SetDefaultValue` from `SetDefaultValueFactory` and added tests to clarify the behaviors. Do you think this helps with the ambiguity?

These APIs still feel fairly provisional though, and a few issues (#581, #704, #723) have pointed toward the need for a more holistic approach to default values, custom conversion, and validation. I won't try to address this here but I suspect it might result in  `SetDefaultValue` and `SetDefaultValueFactory` becoming obsolete.